### PR TITLE
 [Tooling] Integrate Keybase w/ CLI (Issue #484 )

### LINKED ISSUE #484: [Tooling] Integrate Keybase w/ CLI
## Objective

Integrate the Keybase with the CLI to use keys from the keybase.

## Origin Document

In #459, the first version of the keybase was introduced as a separate package in the V1 codebase.

In #169, the first version of the CLI was introduced to the V1 codebase.

## Goals

- When sending commands via the CLI (stake, send, unstake, etc...), it should use keys from from the Keybase

## Deliverables

- [ ] Identify all entrypoints in `app/client/cli` where the keybase should be integrated
- [ ] Do a global search for `#150` and integrate all these entrypoint with the keybase (a subset of the above)
- [ ] Documentation on what the default keys in the keybase are (until the CLI is added to support this)

## Non-goals / Non-deliverables

-  Adding new commands to the CLI to manage (create/import/export) keys

---

**Creator**: @jessicadaugherty
**Co-owner**: @Olshansk

### PR DESCRIPTION
**Pull request by h5law**
## Description

This PR addresses the `#150` comments throughout the `app/client/cli` directory and integrates the keybase to be used in place of reading keys from files locally.

The new method has been added to the `debug/keystore.go` file `InitialiseDebugKeybase()` which reads the 999 validator private keys from the `build/localnet/manifests/private-keys.yaml` file and stores them in the keybase. This is only run when there are less than 999 keys in the keybase. To speed this up `BadgerDB`'s `WriteBatch()` function is being used - however due to the encryption/JSON encoding and then Gob encoding steps of adding private keys to the keybase this takes a few seconds when ran for the first time.

This PR does not add any endpoints to interact with the keybase from the CLI this is instead tracked in the issue #489 

## Issue

Fixes #484 

## Type of change

Please mark the relevant option(s):

- [x] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other 

## List of changes

- Add logic to initialise keybase with 999 validator keys
- Update `app/client/cli` to use the keybase

## Testing

- [x] `make develop_test`
- [x] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

[localnet_with_integration.webm](https://user-images.githubusercontent.com/53987565/217543043-21c35c3e-8389-44a5-968a-d3afb6cb8dd9.webm)

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [x] I have updated the corresponding CHANGELOG

### If Applicable Checklist

- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

File: app/client/keybase/keystore.go
Comment (Olshansk) [inline app/client/keybase/keystore.go:106]: Comment that hint and pwd are empty because these are debug keys

File: app/client/keybase/keystore.go
Comment (Olshansk) [inline app/client/keybase/keystore.go:103]: good use of wb

File: app/client/keybase/keystore.go
Comment (Olshansk) [inline app/client/keybase/keystore.go:62]: I'd prefer to refactor this a bit. We are doing two things here:

1. Initialising a debug keybase for DEV purposes
2. Retrieving a keybase for PROD purposes

My request would be to move `InitialiseDebugKeybase` as a helper inside one of the `init` functions so it's a one-time check. Simply refer to that keybase in the normal code flow.

**Review by Olshansk — CHANGES_REQUESTED** [review]
@h5law Two major comments:

1. I think there's some opportunity to reduce the amount of code duplication
2. To make the code clearer for other devs (and make future changes simpler), we should deliniate dev & prod functionality.

I gave some suggestions/guidelines in the comments but nothing too specific, so lmk if I should provide additional details.

File: app/client/keybase/keystore.go
Comment (h5law) [reply app/client/keybase/keystore.go]: @Olshansk I have added a `debug` folder inside the keybase and am initialising the DB at `$HOME/.pocket/keys` for the debug DB so this can be done in the `init()` function otherwise there would be no way to use `init()` with a path that could change.

File: app/client/cli/gov.go
Comment (h5law) [reply app/client/cli/gov.go:89]: @Olshansk this command seems tightly coupled to be specific to the actor case as the `actorCmdDef` option part of the `applySubcommandOptions` function which uses the `attachPwdFlagToSubcommands` relies on the actor type being defined - it might be better to refactor this logic to be generally applicable to arbitrary commands

Comment (h5law) [reply app/client/cli/gov.go]: I changed the `applySubcommandOptions` to recieve only the options part of the `actorCmdDef` type so this can be used with the result of `attachPwdFlagToSubcommands` directly

File: app/client/cli/account.go
Comment (Olshansk) [inline app/client/cli/account.go:13]: Can you look into `cli/actor.go` and follow the same pattern w.r.t attaching the password params: 
1. Use options to set a pattern in case we need to add others
2. Don't embed it deep inside of `accountCommands` so its easier to find

```go
func init() {
	rootCmd.AddCommand(NewActorCommands(attachPwdFlagToSubcommands())...)

	rawChainCleanupRegex = regexp.MustCompile(rawChainCleanupExpr)

	oneMillion = big.NewInt(1e6)
}
```

File: app/client/cli/account.go
Comment (Olshansk) [reply app/client/cli/account.go:55]: Just TOL: Not going to ask for it here cause I think it's overkill, but unpacking the args into vars near the top could also help with clarity. Only do this if you personally feel it adds value.

File: app/client/cli/actor.go
Comment (Olshansk) [inline app/client/cli/actor.go:110]:
```suggestion
			fromAddr := crypto.AddressFromString(fromAddrHex)
```

Comment (Olshansk) [reply app/client/cli/actor.go:110]: Ditto below

File: app/client/cli/gov.go
Comment (Olshansk) [inline app/client/cli/gov.go:90]: Can you move this into `init` similar to the pattern in `actor.go`

File: app/client/cli/gov.go
Comment (Olshansk) [reply app/client/cli/gov.go]: Change to `applySubcommandOptions` makes sense to me

File: app/client/keybase/debug/keystore.go
Comment (Olshansk) [inline app/client/keybase/debug/keystore.go:22]: s/DebugKeybasePath/debugKeybasePath
```suggestion
         // TODO: Allow users to override this value via `datadir` flag
	debugKeybasePath string
```

File: app/client/keybase/debug/keystore.go
Comment (Olshansk) [inline app/client/keybase/debug/keystore.go:43]:
```suggestion
// Struct to process the yaml file of pre-generated private-keys
```

File: app/client/keybase/debug/keystore.go
Comment (Olshansk) [inline app/client/keybase/debug/keystore.go:52]:
```suggestion
// Creates/Opens the DB and initialises the keys from the pre-generated YAML file of private keys
```

File: app/client/keybase/debug/keystore.go
Comment (Olshansk) [inline app/client/keybase/debug/keystore.go:58]: Move to a const

File: app/client/cli/account.go
Comment (h5law) [reply app/client/cli/account.go:13]: @Olshansk without adding excess code the `gov.go` and `account.go` files don't need I feel this is the best solution:

```go
func NewAccountCommand() *cobra.Command {
	cmd := &cobra.Command{
		Use:     "Account",
		Short:   "Account specific commands",
		Aliases: []string{"account"},
		Args:    cobra.ExactArgs(0),
	}

	cmds := accountCommands()
	applySubcommandOptions(cmds, attachPwdFlagToSubcommands())
	cmd.AddCommand(cmds...)

	return cmd
}
```

similar with the gov one.

The way the actor one works is quite different from the others as it uses the `actorCmdDef` struct which is unique to this CLI endpoint - the others don't have something like this and its definitely not needed (as the code is now) for them either so they shouldn't go in the `init()` function as it would require changes that I don't really think are of any benefit to the structure of the other endpoints. Happy to do it if you think this move is not enough though.

File: app/client/keybase/keystore.go
Comment (Olshansk) [reply app/client/keybase/keystore.go]: It's a step in the right direction, but was hoping to take it further.

At the moment, every function in our PROD CLI starts with a `NewDebugKeybase()` and I think we can make it really prod-ready in this PR.

```
		RunE: func(cmd *cobra.Command, args []string) error {
			// Open the debug keybase at $HOME/.pocket/keys
			kb, err := debug.NewDebugKeybase()
```

The `init` function gets called when we import something. What I'm thinking is:

1. Create a file that contains an anonymous import of the debug package
2. By importing it, we immediately load the debug DB if its empty (we can exclude this from production binaries)
3. Replace the `kb` assignment with a prod ready assignment

File: app/client/keybase/debug/keystore.go
Comment (Olshansk) [inline app/client/keybase/debug/keystore.go:53]: This is in the path name and method name, can be removed now.

File: app/client/keybase/keybase.go
Comment (Olshansk) [inline app/client/keybase/keybase.go:11]:
```suggestion
        // Debug
	// TECHDEBT: Exposed only so we can leverage badger's batch writing features when populating the development keybase
	GetBadgerDB() *badger.DB
```

**Review by Olshansk — CHANGES_REQUESTED** [review]
@h5law Would like to iterate on this a bit (see my comments) so we have a "prod ready" integration between the CLI and the keybase, and not need to revisit this in the future.

Lmk if I can provide any more details.

**Comment by h5law** [comment]
@Olshansk The CLI endpoints are now using "prod ready" code - however the debug keybase will only initialise to `$HOME/.pocket/keys` and I am unsure of a way to make this changeable with `--data_dir` as it is ran from the `init()` function - unless the datadir is binded with viper and used in the debug package?

Regarding the anonymous import I think I have followed what you asked for - not sure what you meant about making an empty file so I just did an anonymous import of the debug package in the `cmd.go` file as this works how you want it to.

File: app/client/cli/account.go
Comment (Olshansk) [reply app/client/cli/account.go:13]: Great explanation and thanks for putting so much thought into it. The current solution LGTM

File: app/client/keybase/debug/keystore.go
Comment (Olshansk) [inline app/client/keybase/debug/keystore.go:18]: Rename to `numValidators` or `validatorCount` and add a comment `// The number of pre-generated keys in privateKeysYamlFile`

File: app/client/cli/actor.go
Comment (Olshansk) [inline app/client/cli/actor.go:100]: Not a debug anymore?

File: app/client/cli/cmd.go
Comment (Olshansk) [inline app/client/cli/cmd.go:8]:
```suggestion
        // NOTE: Imported for debug purposes in order to populate the keybase with the pre-generated keys
	_ "github.com/pokt-network/pocket/app/client/keybase/debug"
```

**Review by Olshansk — APPROVED** [review]
@h5law Leaving a couple of minor last-minute nits but approving so you can merge right after.

Thanks for bearing on the back & forth, but I honestly think that final outcome is really good and much closer to being prod ready.

File: app/client/cli/actor.go
Comment (h5law) [reply app/client/cli/actor.go:100]: @Olshansk no this now uses the prod keybase - however while we do the anonymous import the prod keybase at `$HOME/.pocket/keys` is being initialised with the 999 keys. Remove the import and it has no default keys and its using the normal prod ready DB

Comment (Olshansk) [reply app/client/cli/actor.go:100]: 💯

**Pull request merged**
